### PR TITLE
chore: add wds datepicker widget

### PR DESCRIPTION
**Pull request by jsartisan**
CleanShot 2024-11-26 at 15 56 15

/ok-to-test tags="@tag.Anvil"

## Release Notes

- **New Features**
  - Introduced the WDS Date Picker Widget, enhancing date selection capabilities within the UI.
  - Added configuration options for widget size, visibility, and autocomplete behavior.
  - Implemented comprehensive validation for date input, ensuring accurate user selections.
  - Expanded widget collection to include the new WDS Date Picker Widget.
  - Introduced new constants for date format options, facilitating diverse formatting choices.
  - Added support for a new "Date" input type, enhancing input widget configurability.

- **Documentation**
  - Updated property pane configurations to include detailed settings for date format, validation, and event handling.

- **Bug Fixes**
  - Improved handling of derived properties to ensure proper context during widget interactions.

These updates collectively improve user experience and flexibility in date selection within the application.

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/12083004714>
> Commit: b17348e03db911501970d2c8a59c4fea30a175e1
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=12083004714&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Anvil`
> Spec:
> <hr>Fri, 29 Nov 2024 10:44:39 UTC

### CODE REVIEW

File: app/client/src/widgets/WidgetUtils.ts [hunk 1]
```diff
@@ -989,3 +989,48 @@ export const checkForOnClick = (e: React.MouseEvent<HTMLElement>) => {
...
+ * ```js
+ * {
+ *  isValidDate: "{{ this.value === 1 }}"
+ * }
+ * ```
+ *
+ * Main rule to remember is don't deconstruct the props like `const { value } = props;` in the derived property function.
+ * Directly access props like `props.value`
+ */
+export function parseDerivedProperties(propertyFns: Record<string, unknown>) {
```
Comment (jsartisan) [inline app/client/src/widgets/WidgetUtils.ts:1014]: This is a new piece of code. I need to use it for all widgets as well that I am doing here - https://github.com/appsmithorg/appsmith/pull/37390

**Review by KelvinOm — COMMENTED** [review]
@ichik Should this widget morph to other input type widgets?

**Review by KelvinOm — APPROVED** [review]

**Comment by ichik** [comment]
> @ichik Should this widget morph to other input type widgets?

@KelvinOm I would think so.

File: app/client/src/modules/ui-builder/ui/wds/WDSDatePickerWidget/config/metaConfig.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { WIDGET_TAGS } from "constants/WidgetConstants";
+
+export const metaConfig = {
+  name: "DatePicker",
+  tags: [WIDGET_TAGS.INPUTS],
+  needsMeta: true,
+  searchTags: [
+    "date",
```
[the remark this answers is not in the document]
Comment (ichik) [reply app/client/src/modules/ui-builder/ui/wds/WDSDatePickerWidget/config/metaConfig.ts:9]: > @ichik Could you check this please?

@KelvinOm @jsartisan pushed an update to metaconfig.

**Comment by jsartisan** [comment]
@KelvinOm I have added the input morphing as well in this PR.

File: app/client/src/modules/ui-builder/ui/wds/WDSBaseInputWidget/constants.ts [hunk 3]
```diff
@@ -6,6 +6,7 @@ export const INPUT_TYPES = {
   PASSWORD: "PASSWORD",
   PHONE_NUMBER: "PHONE_NUMBER",
   MULTI_LINE_TEXT: "MULTI_LINE_TEXT",
+  DATE: "DATE",
```
Comment (KelvinOm) [inline app/client/src/modules/ui-builder/ui/wds/WDSBaseInputWidget/constants.ts:9]: After the drop, date picker widget has an empty Data type. Could you please fix this?

Снимок экрана 2024-11-29 в 11 38 37

**Pull request merged**